projects/canopy/src/lib/utils/dynamic-style.service.ts

```diff
@@ -0,0 +1,70 @@
+import { Injectable } from '@angular/core';
+
+import { BreakpointValues } from '../shared/breakpoints.interface';
+
+@Injectable({
+  providedIn: 'root',
+})
+/**
+ * Dynamically add styles to a <style> tag in <head>
+ */
+export class DyanmicStyleService {
+  styleTag = null;
+  selectors = [];
+  mediaQueries = {};
+
+  private createStyleTag(): void {
+    this.styleTag = document.createElement('style');
+    this.styleTag.type = 'text/css';
+    document.getElementsByTagName('head')[0].appendChild(this.styleTag);
+  }
+
+  private addMediaQuery(breakpoint: BreakpointValues): void {
+    if (!this.styleTag) {
+      this.createStyleTag();
+    }
+    if (this.mediaQueries[breakpoint]) {
+      return;
+    }
+    this.mediaQueries[breakpoint] = [];
+    this.styleTag.innerHTML += `@media(min-width:${breakpoint}){}`;
+  }
+
+  addStyle(selector: string, style: string): void {
+    if (!this.styleTag) {
+      this.createStyleTag();
+    }
+    if (this.selectors.includes(selector)) {
+      return;
+    }
+    this.selectors.push(selector);
+    const styleString = `.${selector}{${style}!important}`;
+    this.styleTag.innerHTML += styleString;
+  }
+
+  addStyleToMediaQuery(
+    selector: string,
+    style: string,
+    breakpoint: BreakpointValues,
+  ): void {
+    if (!this.styleTag) {
+      this.createStyleTag();
+    }
+    if (!this.mediaQueries[breakpoint]) {
+      this.addMediaQuery(breakpoint);
+    }
+    if (this.mediaQueries[breakpoint].includes(selector)) {
+      return;
+    }
+
+    this.mediaQueries[breakpoint].push(selector);
+
+    const styleString = `.${selector}{${style}!important}`;
+
+    const innerHTML = this.styleTag.innerHTML;
+    const mqIndex = innerHTML.indexOf(`@media(min-width:${breakpoint}){`) + 1;
+    const htmlArr = [...innerHTML];
+    htmlArr.splice(mqIndex, 0, styleString);
+    this.styleTag.innerHTML = htmlArr.join('');
```

Comment:
I wonder whether changing the innerHTML on multiple occasions could trigger noticeable repaints 🤔
If so, could we just change it once?
Maybe even via an `addRules` method for multiple ones, to cut down the repaint?

Reply:
I've updated it now so that it only updates the innerHTML once after each set of rules or media queries are created, rather than each time. This is noticeable when the responsive spacing object is used, as it may contain two or three rules, with two or three media queries, which now are only added once they are all created. That brings it from 3 to 6 innerHTML changes to only 1.